gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-account/src/main/java/io/gravitee/am/gateway/handler/account/resources/AccountFactorsEndpointHandler.java

```diff
@@ -592,10 +593,12 @@ private List<String> getUserRecoveryCodes(User user){
     }
 
     private Factor getClientRecoveryCodeFactor(Client client) {
-        for (String factorId : client.getFactors()) {
-            Factor factor = factorManager.getFactor(factorId);
-            if (isRecoveryCodeFactor(factor)) {
-                return factor;
+        if (client.getFactorSettings() != null && client.getFactorSettings().getApplicationFactors() != null) {
+            for (ApplicationFactorSettings factorSettings : client.getFactorSettings().getApplicationFactors()) {
```

Comment:
suggestion: 
it could be better to use, like this:
Optional.of(client.getFactorSettings()).map(i -> i.getApplicationFactors()).map(factorSettings -> factorSettings.stream().filter(...).findFirst().orElse(null)